R/fread.R

```diff
@@ -164,29 +169,61 @@ autostart=NA)
   if (check.names) {
     setattr(ans, 'names', make.names(names(ans), unique=TRUE))
   }
-  cols = NULL
-  if (stringsAsFactors)
-    cols = which(vapply(ans, is.character, TRUE))
-  else if (length(colClasses)) {
-    if (is.list(colClasses) && "factor" %chin% names(colClasses))
-      cols = colClasses[["factor"]]
-    else if (is.character(colClasses) && "factor" %chin% colClasses)
-      cols = which(colClasses=="factor")
-  }
-  setfactor(ans, cols, verbose)
-  # 2007: is.missing is not correct since default value of select is NULL
+
+  colClassesAs = attr(ans, "colClassesAs")   # should only be present if one or more are != ""
+  for (j in which(colClassesAs!="")) {       # # 1634
+    v <- .subset2(ans, j)
+    new_class = colClassesAs[j]
+    new_v <- tryCatch({    # different to read.csv; i.e. won't error if a column won't coerce (fallback with warning instead)
+      switch(new_class,
+             "factor" = as_factor(v),
+             "complex" = as.complex(v),
+             "raw" = as_raw(v),  # Internal implementation
+             "Date" = as.Date(v),
+             "POSIXct" = as.POSIXct(v),
+             # finally:
+             methods::as(v, new_class))
+      },
+      warning = fun <- function(e) {
+        etype = if (inherits(e,"error")) "error" else "warning"
+        warning(sprintf("Column '%s' was set by colClasses to be '%s' but fread encountered the following %s:\n\t%s\nso the column has been left as type '%s'",
```

Comment:
Construct without `sprintf` (I don't think I have write access to Hugh's branch...)

Reply:
I don't mind either way, but it's ok with `sprintf` isn't it?

Reply:
I took the cue from [WRE Section 1.7](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Diagnostic-messages):

> Try not to split up messages into small pieces. In C error messages use a single format string containing all English words in the messages.
> In R error messages do not construct a message with `paste` (such messages will not be translated) but via multiple arguments to `stop` or `warning`, or via `gettextf`.

I already purged other instances of this construction elsewhere in the code, I can't find the PR right now...

Reply:
As it's merged now, follow up PRs are good.  I double-checked Hugh is a project member so he can create branches too.

Reply:
yep, working on it 👍